### PR TITLE
Add new command line argument - array

### PR DESCRIPTION
**Pull request by Antondomashnev**
Hi there, this PR brings the first step of making [skip schemes with Carthage](https://github.com/Carthage/Carthage/issues/1227) possible. Next step I'll try to actually implement scheme filtering in Carthage project itself.

Looking forward to code review and feedback 😉

### CODE REVIEW

**Comment by Antondomashnev** [comment]
Hi @ikesyo @jspahrsummers would you mind to check the PR, thanks 👍

**Comment by Antondomashnev** [comment]
@mdiep thanks for the review. If the Carthage already has something and I don't have to involve Commandant to implement the same idea - that is even better 👍 I'll take a look on it and will follow up here if I have any questions.

**Comment by mdiep** [comment]
Carthage's isn't generic. So I think an addition here would be valuable. But it should work with an `enum` like Carthage's.

**Comment by Antondomashnev** [comment]
@mdiep hi, I've read once again the Carthage code with `--platform` argument parsing. Could you please clarify what could be the similarity between platform and scheme?
I mean, currently, right now there are predefined platforms and you can control the user input while checking if it contains only accepted platforms, however, with schemes I don't see this possibility, unless it parses the project info first to check if these schemes exist or not.

The only possibility to make it less generic for Carthage I see is to basically copy paste the solution from this PR info Carthage and rename it to the array of schemes, while they still will be the string that basically an `ArgumentProtocol `.

@mdiep probably I'm misunderstanding some architectural decisions in Carthage implementation. Could you please clarify a bit more your opinion and suggested implementation?

**Comment by mdiep** [comment]
TBH I don't think we'd want a `--schemes` option for Carthage. Any configuration like that would _need_ to happen at the project level—not at the CLI.

But I do still think this would be a useful option. Looking at the actual code a bit more, I think something like this would probably be suitable:

```swift
public func <| <T: ArgumentType, ClientError>(mode: CommandMode, option: Option<[T]>) -> Result<T?, CommandantError<ClientError>> {
	let key = option.key
	switch mode {
	case let .Arguments(arguments):
		var stringValue: String?
		switch arguments.consumeValueForKey(key) {
		case let .Success(value):
			stringValue = value

		case let .Failure(error):
			switch error {
			case let .UsageError(description):
				return .Failure(.UsageError(description: description))

			case .CommandError:
				fatalError("CommandError should be impossible when parameterized over NoError")
			}
		}

		if let stringValue = stringValue {
			let values: [T] = []
			let components = stringValue.components(separatedBy: CharacterSet(charactersIn: ", ")

			for stringValue in components {
				if let value = T.fromString(stringValue) {
					values.append(value)
				} else {
					let description = "Invalid value for '--\(key)': \(stringValue)"
					return .Failure(.UsageError(description: description))	
				}
			}
			return .Success(values)
		} else {
			return .Success(option.defaultValue)
		}

	case .Usage:
		return .Failure(informativeUsageError(option))
	}
}
```

That would still be workable with an `Option<[String]>`. (Please forgive the mix of Swift 2 and Swift 3 in that snippet. But it should give you something to work from.)

**Comment by Antondomashnev** [comment]
@mdiep thanks for the snippet 👍 , will read it carefully.

> TBH I don't think we'd want a --schemes option for Carthage

This idea came up only after my personal needs and the needs of some other people from the related [PR](https://github.com/Carthage/Carthage/issues/1227). Is there any downsides from your perspective to have scheme option for Carthage to build only specific schemes?

**Comment by Antondomashnev** [comment]
@mdiep I've just tried to replicate your idea, but I think you missed the fact that the generic `T` is used to create an argument. So in the case of `T: [String]` the code expects that `[String]` conforms to `ArgumentProtocol`, but apparently only `String` conforms to it. In my proposal, it's exactly the case why I introduced a new `Argument` - to be able the "array" to produce a value from `String`.

**Comment by mdiep** [comment]
> So in the case of T: [String] the code expects that [String] conforms to ArgumentProtocol, but apparently only String conforms to it.

The point of the added operation is that it handles the case of `[T]` where `T: ArgumentProtocol`. The array isn't part of the `T`: the array contains `T`s.

```swift
public func <| <T: ArgumentType, ClientError>(mode: CommandMode, option: Option<[T]>) -> Result<T?, CommandantError<ClientError>> {
```

**Comment by Antondomashnev** [comment]
@mdiep thanks, now it looks much clearer 👍 
Let's see the Travis result. If it's green -> ready to be reviewed.

File: Sources/Commandant/Option.swift [hunk 1]
```diff
@@ -193,6 +193,47 @@ public func <| <T: ArgumentProtocol, ClientError>(mode: CommandMode, option: Opt
...
+				
+			case .commandError:
+				fatalError("CommandError should be impossible when parameterized over NoError")
+			}
+		}
+	
+		guard let unwrappedStringValue = stringValue else {
+			return .success(option.defaultValue)
+		}
+		let components = unwrappedStringValue.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
```
Comment (mdiep) [inline Sources/Commandant/Option.swift]: This should probably split on both `,` and ` `. (That's the existing behavior in Carthage for `--platform`.)

File: Sources/Commandant/Option.swift [hunk 2]
```diff
@@ -193,6 +193,47 @@ public func <| <T: ArgumentProtocol, ClientError>(mode: CommandMode, option: Opt
 	}
 }
 
+/// Evaluates the given option in the given mode.
+///
+/// If parsing command line arguments, and no value was specified on the command
+/// line, `nil` is used.
+public func <| <T: ArgumentProtocol, ClientError>(mode: CommandMode, option: Option<[T]?>) -> Result<[T]?, CommandantError<ClientError>> {
```
Comment (mdiep) [inline Sources/Commandant/Option.swift]: There should probably be a non-nil version of this as well?

File: Tests/CommandantTests/ArgumentArraySpec.swift [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+//					var argumentsArray: ArgumentsArray?
+//					
+//					beforeEach {
+//						argumentsArray = ArgumentsArray.from(string: arguments)
+//					}
+//					
+//					it("returns valid ArgumentsArray") {
+//						expect(argumentsArray?.arguments).to(equal(["one", "two", "three"]))
+//					}
+//				}
```
Comment (mdiep) [inline Tests/CommandantTests/ArgumentArraySpec.swift]: This file needs to be removed.

**Comment by mdiep** [comment]
Yeah, I think this is looking a lot better. 👏

File: Sources/Commandant/Option.swift (hunk 1 above)
Comment (Antondomashnev) [inline Sources/Commandant/Option.swift]: I'm not sure if I agree. Could you please explain your thoughts why `, ` could be better than splitting with `,` and after trimming all whitespaces. I think my solution could handle the format mistakes such as ` , ` or ` ,`.

File: Tests/CommandantTests/ArgumentArraySpec.swift (hunk 3 above)
Comment (Antondomashnev) [inline Tests/CommandantTests/ArgumentArraySpec.swift]: Nice catch thanks 😄 Removed in 449fcc1

File: Sources/Commandant/Option.swift (hunk 2 above)
Comment (Antondomashnev) [inline Sources/Commandant/Option.swift]: Yep, makes sense. I've created one in 449fcc1. I was following what was existed for non-array version. Seems reasonable to me.

File: Sources/Commandant/Option.swift (hunk 1 above)
Comment (mdiep) [inline Sources/Commandant/Option.swift]: It's not splitting on `", "`—it's splitting on either `,` or ` `. So `mac,ios`, `mac, ios`, and `mac ios` all mean the same thing.

**Comment by mdiep** [comment]
Can you remove the `.DS_Store` this adds to the root directory? Once that's done, this is ready to go!

File: Sources/Commandant/Option.swift [hunk 4]
```diff
@@ -84,6 +84,15 @@ extension Option: CustomStringConvertible {
 	}
 }
 
+extension String {
+	/// Split the string into substrings separated by the given separators.
+	internal func split(by separators: [Character] = [ ",", " " ], allowEmptySlices: Bool = true) -> [String] {
+		return characters
+				.split(omittingEmptySubsequences: allowEmptySlices, whereSeparator: separators.contains)
```
Comment (ikesyo) [inline Sources/Commandant/Option.swift]: `allowEmptySlices` is the opposite of `omittingEmptySubsequences`.

**Comment by Antondomashnev** [comment]
@mdiep @ikesyo this is a bit of offtopic of PR, but could you share your thoughts why you don't see a need to be able to select a scheme on Carthage? 😄

**Pull request merged**